datastore/mongodb/filter_outputs.go

```diff
@@ -30,8 +31,50 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
 func (c *Client) UpdateFilterOutput(ctx context.Context, f *model.FilterOutput) error {
 	col := c.collections.filterOutputs
 
-	if _, err := c.conn.Collection(col.name).UpsertById(ctx, f.ID, bson.M{"$set": f}); err != nil {
-		return errors.Wrap(err, "failed to upsert filter")
+	var docs []model.FilterOutput
+	sc := bson.M{"id": f.ID}
+
+	var err error
+	var num int
+
+	if num, err = c.conn.Collection(col.name).Find(ctx, sc, &docs, mongodb.Limit(1)); err != nil {
+		return errors.Wrap(err, "failed to update filter outputs")
+	}
+	if num < 1 {
+		err := &er{
+			err: errors.Errorf("failed to find filter output"),
+		}
+		err.notFound = true
+		return err
+	}
+
+	//a record with stat 'completed' can't be updated further
+	if docs[0].State == model.COMPLETED {
+		err := &er{
```

Comment:
No need to create the error before returning it here
```suggestion
		return &er{
		  err: errors.Errorf(`filter output is already in "completed" state`),
		  forbidden: true,
		}
		
```